app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt

```diff
@@ -0,0 +1,220 @@
+import org.oppia.android.domain.hintsandsolution.HintsAndSolutionProdModule
+import org.oppia.android.domain.onboarding.ExpirationMetaDataRetrieverModule
+import org.oppia.android.domain.oppialogger.LogStorageModule
+import org.oppia.android.domain.oppialogger.loguploader.LogUploadWorkerModule
+import org.oppia.android.domain.platformparameter.PlatformParameterModule
+import org.oppia.android.domain.platformparameter.PlatformParameterSingletonModule
+import org.oppia.android.domain.question.QuestionModule
+import org.oppia.android.domain.topic.PrimeTopicAssetsControllerModule
+import org.oppia.android.domain.workmanager.WorkManagerConfigurationModule
+import org.oppia.android.testing.TestImageLoaderModule
+import org.oppia.android.testing.TestLogReportingModule
+import org.oppia.android.testing.junit.InitializeDefaultLocaleRule
+import org.oppia.android.testing.robolectric.RobolectricModule
+import org.oppia.android.testing.threading.TestDispatcherModule
+import org.oppia.android.testing.time.FakeOppiaClock
+import org.oppia.android.testing.time.FakeOppiaClockModule
+import org.oppia.android.util.accessibility.AccessibilityTestModule
+import org.oppia.android.util.caching.AssetModule
+import org.oppia.android.util.caching.testing.CachingTestModule
+import org.oppia.android.util.gcsresource.GcsResourceModule
+import org.oppia.android.util.locale.LocaleProdModule
+import org.oppia.android.util.logging.LoggerModule
+import org.oppia.android.util.logging.firebase.FirebaseLogUploaderModule
+import org.oppia.android.util.networking.NetworkConnectionDebugUtilModule
+import org.oppia.android.util.networking.NetworkConnectionUtilDebugModule
+import org.oppia.android.util.parser.html.HtmlParserEntityTypeModule
+import org.oppia.android.util.parser.image.ImageParsingModule
+import org.robolectric.annotation.Config
+import org.robolectric.annotation.LooperMode
+import javax.inject.Inject
+import javax.inject.Singleton
+
+/** Tests for [TextViewBindingAdapters]. */
+@RunWith(AndroidJUnit4::class)
+@LooperMode(LooperMode.Mode.PAUSED)
+@Config(
+  application = TextViewBindingAdaptersTest.TestApplication::class,
+  qualifiers = "port-xxhdpi"
+)
+
+class TextViewBindingAdaptersTest {
+  @get:Rule
+  val initializeDefaultLocaleRule = InitializeDefaultLocaleRule()
+
+  @get:Rule
+  var activityRule: ActivityScenarioRule<TextViewBindingAdaptersTestActivity> =
+    ActivityScenarioRule(
+      Intent(
+        ApplicationProvider.getApplicationContext(),
+        TextViewBindingAdaptersTestActivity::class.java
+      )
+    )
+
+  @Before
+  fun setUp() {
+    setUpTestApplicationComponent()
+    Intents.init()
+  }
+
+  @After
+  fun tearDown() {
+    Intents.release()
+  }
+
+  @Inject
+  lateinit var fakeOppiaClock: FakeOppiaClock
+
+  // Time: Wed Apr 24 2019 08:22:00
+  private val MORNING_TIMESTAMP = 1556094120000
+
+  @Test
+  fun testTextViewBindingAdapters_ltrIsEnabled_port_profileDataTextIsCorrect() {
```

Comment:
These tests cover a couple of the happy path cases, but ought to cover all other cases that the production code includes.

One way to do this: you can analyze the production code to determine different cases of input that would hit each branch of the implementation. Then, consider what behaviors would result in those inputs, and then add tests for each behavior you come up with. In this case, you've only add one behavior for setProfileDataText (setting it to the same timestamp as the current time).

Reply:
The production code for setProfileDataText is https://github.com/oppia/oppia-android/blob/b9d4b13af8c880f5fed8099194f26f9b3f52f0c3/app/src/main/java/org/oppia/android/app/databinding/TextViewBindingAdapters.java#L26. The way to think about this is any if/when blocks and loops indicate multiple conditions that should have corresponding tests. Sometimes methods without control structures can also vary based on input (such as selecting a different string). We should test all the cases we can think of for these methods.

Reply:
Basically `setProfileDataText ` is only responsible for showing "Created on ....." text in textview, which means we are just sending the timestamp as long and it is converting it as string. Now whatever timestamp we send it will convert it to string, that timestamp value can be `0` or `1556094120000` and it will just covert it to either 

`Created on 1 Jan 1970`, or `Created on April 24, 2019`. So I am not sure what other test we can add. 

Are you thinking of other locales like arabic?

Reply:
I think testing setProfileDataText for two different inputs is probably sufficient (to demonstrate that the variable part of the string is working as expected). We don't need to test multiple locales here since there's nothing locale-specific.

setProfileLastVisitedText has a lot more behaviors to test, though, since the label that's computed has a bunch of different potential cases. I see that method as being the one that will have the most tests corresponding to it.

Reply:
Added another case of `TIMESTAMP =0`.

Reply:
So I think checking `setProfileLastVisitedText` for `just now`, `x minutes ago`, `x hours ago` and `x days ago` will be sufficient?

Reply:
@BenHenning @rt4914 I have added other test cases for possible outputs. Do I need to structure them?

Reply:
Thanks @rishidyno! I think this is on the right track.

In terms of testing coverage, I think you did a really good job covering all of the "this should work as expected" cases. The three that I think are also worth covering are when the provided timestamp is in the future (i.e. bigger than the current time), when it's zero, and when it's negative. These are all cases handled by the implementation, so we should verify that they work as expected (as per the adapter's contract).

Beyond that, you should structure each behavior being checked in its own test rather than combining them all into one.

Reply:
@BenHenning I have separated all test cases and added few more. PTAL.

Reply:
Left some follow-up comments, so resolving this now.